File: algorithms/js/romanToInteger/romanToInteger.js
// Source : https://leetcode-cn.com/problems/roman-to-integer/
// Author : Tian Zhuo
// Date   : 2018-11-17
/**
 * @param {string} s
 * @return {number}
 */
var romanToInt = function(s) {
    var romanMap = {
        'I': 1,
        'V': 5,
        'X': 10,
        'L': 50,
        'C': 100,
        'D': 500,
        'M': 1000
    }
    if(s.length == 1) return romanMap[s];
    var romanArr = s.split('');
    var res = 0;
    for (let i = 0; i < romanArr.length; i++) {
        if(romanMap[romanArr[i]] < romanMap[romanArr[i+1]]){
            res = res + romanMap[romanArr[i+1]] - romanMap[romanArr[i]];
            i++
        }else{
            res = res + romanMap[romanArr[i]];
        }
    }
    return res;
};

/**
 * @param {string} s
 * @return {number}
 */

let romanToInt1 = function(s) {
    let map = {I: 1, V: 5, X: 10, L: 50, C: 100, D: 500, M: 1000};
    let res = 0;
    let last = 0;
    for (let i = s.length - 1; i >= 0; i--) {
        let curr = map[s.charAt(i)];   
        if (curr < last) res -= curr;
        else res += curr;
        last = curr;
    }
    return res;
}

File: src/main/java/romanToInteger/RomanToInteger.java
/*
 * @Source: https://leetcode-cn.com/problems/roman-to-integer/
 * @Author: Tian Zhuo
 * @Date: 2018-12-15 10:40:46
 */
package romanToInteger;

import java.util.HashMap;

public class RomanToInteger {
    public int romanToInt(String s) {
        HashMap<Character, Integer> map = new HashMap<Character, Integer>(){{
            put('I', 1);
            put('V', 5);
            put('X', 10);
            put('L', 50);
            put('C', 100);
            put('D', 500);
            put('M', 1000);
        }};
        int i = s.length() - 1;
        int res = 0;
        int prev = 0;
        while(i >= 0) {
            char curr = s.charAt(i);
            int val = map.get(curr);
            if(val < prev) res -= val;
            else res += val;
            prev = val;
            --i;
        }
        return res;
    }
}